### PR TITLE
feat: make Session iterable

### PR DESCRIPTION
**Pull request by yani**
Implemented `\Iterator` interface into `Session` class.  
The `\Iterator` interface is what makes `foreach()` work.   
Also added a testing scenario. 

Closes my own issue: https://github.com/compwright/php-session/issues/10

Still wondering if `Serializable` is worth implementing.

### CODE REVIEW

File: src/Session.php [hunk 1]
```diff
@@ -223,6 +224,31 @@ public function toArray(): array
         return $this->contents ?? [];
     }
 
+    public function rewind(): void
+    {
+        reset($this->contents);
```
Comment (compwright) [inline src/Session.php:234]: Should throw if not initialized. Should fix the test failure.

```suggestion
        if (!$this->isInitialized()) {
            throw new RuntimeException('Session not initialized');
        }
        
        reset($this->contents);
```

File: src/Session.php [hunk 2]
```diff
@@ -223,6 +224,31 @@ public function toArray(): array
...
+    }
+
+    public function current(): mixed
+    {
+        return current($this->contents);
```
Comment (compwright) [inline src/Session.php:244]:
```suggestion
        if (!$this->isInitialized()) {
            throw new RuntimeException('Session not initialized');
        }
        
        return current($this->contents);
```

File: src/Session.php [hunk 3]
```diff
@@ -223,6 +224,31 @@ public function toArray(): array
...
+    }
+
+    public function key(): mixed
+    {
+        return key($this->contents);
```
Comment (compwright) [inline src/Session.php:254]:
```suggestion
        if (!$this->isInitialized()) {
            throw new RuntimeException('Session not initialized');
        }
        
        return key($this->contents);
```

File: src/Session.php [hunk 4]
```diff
@@ -223,6 +224,31 @@ public function toArray(): array
...
+    }
+
+    public function next(): void
+    {
+        next($this->contents);
```
Comment (compwright) [inline src/Session.php:264]:
```suggestion
        if (!$this->isInitialized()) {
            throw new RuntimeException('Session not initialized');
        }
        
        next($this->contents);
```

File: src/Session.php [hunk 5]
```diff
@@ -223,6 +224,31 @@ public function toArray(): array
...
+    }
+
+    public function valid(): bool
+    {
+        return key($this->contents) !== null;
```
Comment (compwright) [inline src/Session.php:274]:
```suggestion
        if (!$this->isInitialized()) {
            throw new RuntimeException('Session not initialized');
        }
        
        return key($this->contents) !== null;
```

**Review by compwright — CHANGES_REQUESTED** [review]
Well done! Thanks for including the test case. We have some test failures, I think it's from the missing guard statements.

**Comment by compwright** [comment]
We don't need Serializable, there's nothing in Session that would require special handling. It's only needed when serializing objects that contain things needing special handling like resources. We have a simple array, and so Session can be serialized just fine as-is.

**Comment by yani** [comment]
```
Error: Parameter #1 $array of function reset expects array|object, array|null given.
Error: Parameter #1 $array of function current expects array|object, array|null given.
Error: Parameter #1 $array of function key expects array|object, array|null given.
Error: Parameter #1 $array of function next expects array|object, array|null given.
Error: Parameter #1 $array of function key expects array|object, array|null given.
```

Reason is `array|null` on `$contents`:
```
    /**
     * @var mixed[]|null
     */
    protected ?array $contents = null;
```

I think `array|object` doesn't break anything so I'll change it to this and adjust accordingly:
```
protected array $contents = [];
```

**Comment by yani** [comment]
By using empty arrays instead of null, `isInitialized()` might be obsolete.

Should I keep the method, and use a property like: `protected bool $is_initialized = false;`?

That way no changes to the Manager need to be done as it uses the method to determine whether or not to open the Session.  I'm wondering if the Manager **always** opens the Session regardless of flow. If so, it would be better to remove the method.

Another method would be to simply ignore the errors.

**Comment by compwright** [comment]
> I think `array|object` doesn't break anything so I'll change it to this and adjust accordingly:
> 
> ```
> protected array $contents = [];
> ```

Let's not do that, it's uninitialized for a reason.



> ```
> Error: Parameter #1 $array of function key expects array|object, array|null given.
> ```
> 
> Reason is `array|null` on `$contents`

The guard statement will prevent acting on a null $contents, but PHPStan doesn't know that, so in this case I think it's fine to just add `// @phpstan-ignore-next-line` or whatever to suppress the error.

**Comment by yani** [comment]
Updated tests as well

```
  Scenario: Iterate over populated session # features\access.feature:30
    When data exists                       # Compwright\PhpSession\BehaviorTest\AccessContext::dataExists()
    Then data is iterated                  # Compwright\PhpSession\BehaviorTest\AccessContext::iteratorSucceeds()

  Scenario: Iterate over non-populated session # features\access.feature:33
    When data does not exist                   # Compwright\PhpSession\BehaviorTest\AccessContext::dataDoesNotExist()
    Then data is not iterated                  # Compwright\PhpSession\BehaviorTest\AccessContext::iteratorFails()
```

**Review by compwright — APPROVED** [review]

**Pull request merged**